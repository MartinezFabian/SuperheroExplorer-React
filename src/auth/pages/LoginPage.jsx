import { useNavigate } from 'react-router-dom';
import styles from './LoginPage.module.css';
import { useContext, useState } from 'react';
import { AuthContext } from '../context/AuthContext';

export const LoginPage = () => {
  const navigate = useNavigate();
  const { login } = useContext(AuthContext);

  const [formState, setFormState] = useState('');

  const onInputChange = (e) => {
    setFormState(e.target.value);
  };

  const onSignIn = (e) => {
    e.preventDefault();

    const username = formState.trim().split(' ').join('');

    if (username.length === 0) return;

    login(username);

    navigate('/', { replace: true });
  };

  return (
    <main className={styles.login}>
      <h1>Login</h1>

      <form onSubmit={onSignIn} className={styles.login__form}>
        <input
          type="text"
          onChange={onInputChange}
          value={formState}
          placeholder="Username"
          autoComplete="off"
          className={styles.login__input}
        />

        <button type="submit" className={styles.login__btn}>
          Sign in
        </button>

        <div
          className={`${styles.login__alert} animate__animated animate__fadeIn`}
          style={{ display: formState.trim().length === 0 ? '' : 'none' }}
        >
          Type the username...
        </div>
      </form>
    </main>
  );
};
