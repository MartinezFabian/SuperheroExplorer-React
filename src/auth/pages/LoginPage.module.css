.login {
  margin-top: 10rem;
  max-width: 60rem;
  width: 80%;
  margin: 10rem auto 0 auto;
}

.login__btn {
  padding: 1rem 2rem;
  border-radius: 1rem;
  background-color: var(--btn-color);
  border-style: none;
  color: var(--white);
  font-weight: 700;
}

.login__btn:hover {
  cursor: pointer;
}
