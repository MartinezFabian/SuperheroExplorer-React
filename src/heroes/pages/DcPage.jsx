export const DcPage = () => {
  return (
    <>
      <h1>DC</h1>
    </>
  );
};
