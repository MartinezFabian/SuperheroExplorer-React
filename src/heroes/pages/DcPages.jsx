export const DcPages = () => {
  return (
    <>
      <h1>DC</h1>
    </>
  );
};
