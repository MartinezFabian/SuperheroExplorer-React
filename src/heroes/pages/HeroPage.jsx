import { Navigate, useNavigate, useParams } from 'react-router-dom';
import { getHeroById } from '../helpers/getHeroById';

import styles from './HeroPage.module.css';
import { useMemo } from 'react';

export const HeroPage = () => {
  const navigate = useNavigate();

  const { id } = useParams();

  /* memorizar el resultado de la función, el resultado se calculará solo cuando el id cambie. Si 
    el id permanece igual, el valor memorizado se reutilizará */
  const hero = useMemo(() => getHeroById(id), [id]);

  if (!hero) {
    return <Navigate to="/404"></Navigate>;
  }

  const { superhero, publisher, alter_ego, first_appearance, characters } = hero;

  const heroImg = `/assets/img/${id}.jpg`;

  const onNavigateBack = () => {
    if (hero.publisher === 'DC Comics') {
      navigate('/dc');
    } else if (hero.publisher === 'Marvel Comics') {
      navigate('/marvel');
    }
  };

  return (
    <main className={styles.main}>
      <section className={styles.main__section}>
        <img src={heroImg} alt={superhero} className={styles.main__img} />
      </section>

      <section className={styles.information}>
        <h2 className={styles.information__heading}>{superhero}</h2>

        <div>
          <p className={styles.information__text}>
            Alter Ego: <span className={styles.information__span}>{alter_ego}</span>
          </p>
          <p className={styles.information__text}>
            Publisher: <span className={styles.information__span}>{publisher}</span>
          </p>
          <p className={styles.information__text}>
            First Appearance: <span className={styles.information__span}>{first_appearance}</span>
          </p>
        </div>

        <div>
          <h3 className={styles.information__heading}>Characters</h3>
          <p>{characters}</p>
        </div>

        <button onClick={onNavigateBack} className={styles.information__btn}>
          Back
        </button>
      </section>
    </main>
  );
};
