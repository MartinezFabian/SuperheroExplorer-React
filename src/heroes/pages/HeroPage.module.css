.main {
  width: 90%;
  max-width: 120rem;
  margin: 10rem auto;
}

@media (min-width: 468px) {
  .main {
    display: grid;
    grid-template-columns: 1fr 1fr;
    gap: 4rem;
  }
}

.main__img {
  max-width: 100%;
  display: block;
  height: auto;
  object-fit: cover;
  box-shadow: rgba(0, 0, 0, 0.3) 0px 19px 38px, rgba(0, 0, 0, 0.22) 0px 15px 12px;
}

.information {
  margin-top: 1rem;
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.information__heading {
  margin: 0;
}

.information__text {
  font-weight: 700;
}

.information__span {
  font-weight: 400;
}

.information__btn {
  padding: 1rem 2rem;
  border-radius: 1rem;
  background-color: var(--btn-color);
  border-style: none;
  color: var(--white);
  font-weight: 700;
}

.information__btn:hover {
  cursor: pointer;
}

@media (min-width: 468px) {
  .information__btn {
    max-width: 11rem;
  }
}
