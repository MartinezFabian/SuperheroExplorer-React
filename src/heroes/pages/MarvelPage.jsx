import { HeroesGrid } from '../components/HeroesGrid';

import styles from './MarvelPage.module.css';

export const MarvelPage = () => {
  return (
    <main className={styles.container}>
      <h1 className={styles.heading}>Marvel</h1>

      <HeroesGrid publisher="Marvel Comics"></HeroesGrid>
    </main>
  );
};
