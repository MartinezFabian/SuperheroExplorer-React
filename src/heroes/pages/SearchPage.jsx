import { useLocation } from 'react-router-dom';
import { HeroCard } from '../components/HeroCard';
import { HeroSearch } from '../components/HeroSearch';
import { getHeroesByName } from '../helpers/getHeroesByName';
import styles from './SearchPage.module.css';

export const SearchPage = () => {
  // Obtener el objeto location que representa la ubicación actual de la página
  const location = useLocation();

  // Crear un objeto 'searchParams' para manejar los parámetros de búsqueda de la URL
  const searchParams = new URLSearchParams(location.search);

  //Se intenta obtener el valor del parámetro de búsqueda 'q' de la URL
  const q = searchParams.get('q') ? searchParams.get('q') : '';

  const heroes = getHeroesByName(q);

  const showMessage = q === '';
  const showError = heroes.length === 0 && !showMessage;

  return (
    <main className={styles.main}>
      <section className={styles.container}>
        <HeroSearch></HeroSearch>
      </section>

      <section className={styles.results}>
        <div
          className={`${styles.results__alert} ${styles['results__alert--info']} animate__animated animate__fadeIn`}
          style={{ display: showMessage ? '' : 'none' }}
        >
          Type the hero name
        </div>

        <div
          aria-label="alert-not-found"
          className={`${styles.results__alert} animate__animated animate__fadeIn`}
          style={{ display: showError ? '' : 'none' }}
        >
          The hero is not found
        </div>

        <ul className={`${styles.results__grid} animate__animated animate__fadeIn`}>
          {heroes.map((hero) => (
            <HeroCard key={hero.id} {...hero}></HeroCard>
          ))}
        </ul>
      </section>
    </main>
  );
};
