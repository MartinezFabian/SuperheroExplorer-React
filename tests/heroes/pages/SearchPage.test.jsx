import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../../src/heroes/pages/SearchPage';

/* eslint-disable no-undef */
describe('tests in SearchPage.jsx', () => {
  test('must be displayed correctly', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage></SearchPage>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test('superman must be shown and alert not found must have display on none', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=superman']}>
        <SearchPage></SearchPage>
      </MemoryRouter>
    );

    const h3 = screen.getByRole('heading', { level: 3 });
    expect(h3.textContent).toBe('Superman');

    const img = screen.getByRole('img');
    expect(img.src).toContain('/img/dc-superman.jpg');

    const alertNotFound = screen.getByLabelText('alert-not-found');
    expect(alertNotFound.style.display).toBe('none');
  });

  test('must display the alert if it is a non-existent hero', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=nonexistenthero']}>
        <SearchPage></SearchPage>
      </MemoryRouter>
    );

    const alertNotFound = screen.getByLabelText('alert-not-found');
    expect(alertNotFound.style.display).toBe('');
  });
});
