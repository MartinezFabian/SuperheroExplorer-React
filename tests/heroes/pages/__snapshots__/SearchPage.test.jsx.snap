// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in SearchPage.jsx must be displayed correctly 1`] = `
<div>
  <main
    class="main"
  >
    <section
      class="container"
    >
      <form
        class="search"
      >
        <input
          autocomplete="off"
          class="search__input"
          name="searchInput"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="search__btn"
          type="submit"
        >
          <svg
            class="search__icon"
            fill="none"
            stroke="#dfdfdf"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </form>
    </section>
    <section
      class="results"
    >
      <div
        class="results__alert results__alert--info animate__animated animate__fadeIn"
      >
        Type the hero name
      </div>
      <div
        aria-label="alert-not-found"
        class="results__alert animate__animated animate__fadeIn"
        style="display: none;"
      >
        The hero is not found
      </div>
      <ul
        class="results__grid animate__animated animate__fadeIn"
      />
    </section>
  </main>
</div>
`;
